app/helpers/application_helper/toolbar_builder.rb

```diff
@@ -40,14 +40,189 @@ def load_yaml(tb_name)
   end
 
   def build_toolbar(tb_name)
-    tb_hash = tb_name == "custom_buttons_tb" ? build_custom_buttons_toolbar(@record) : load_yaml(tb_name)
+    definition = tb_name == "custom_buttons_tb" ? build_custom_buttons_toolbar(@record) : load_yaml(tb_name)
+    build(definition)
+  end
+
+  def build_select_button(bgi, index)
+    bs_children = false
+    props = {
+      "id"     => bgi[:buttonSelect],
+      "type"   => "buttonSelect",
+      "img"    => img = "#{bgi[:image] ? bgi[:image] : bgi[:buttonSelect]}.png",
+      "imgdis" => img,
+      :icon    => bgi[:icon]
+    }
+    props["title"] = bgi[:title] unless bgi[:title].blank?
+    props["text"]  = bgi[:text]  unless bgi[:text].blank?
+    if bgi[:buttonSelect] == "history_choice" && x_tree_history.length < 2
+      props["enabled"] = false  # Show disabled history button if no history
+    else
+      props["enabled"] = "#{bgi[:enabled]}" unless bgi[:enabled].blank?
+    end
+    props["openAll"] = true # Open/close the button select on click
+
+    if bgi[:buttonSelect] == "chargeback_download_choice" && x_active_tree == :cb_reports_tree &&
+       @report && !@report.contains_records?
+      props["enabled"] = "false"
+      props["title"] = _("No records found for this report")
+    end
+    if bgi[:buttonSelect] == "host_vmdb_choice" && x_active_tree == :old_dialogs_tree && @record && @record[:default]
+      bgi[:items].each do |bgsi|
+        if bgsi[:button] == "old_dialogs_edit"
+          bgsi[:enabled] = 'false'
+          bgsi[:title] = _('Default dialogs cannot be edited')
+        elsif bgsi[:button] == "old_dialogs_delete"
+          bgsi[:enabled] = 'false'
+          bgsi[:title] = _('Default dialogs cannot be removed from the VMDB')
+        end
+      end
+    end
+    if bgi[:buttonSelect] == "orchestration_template_vmdb_choice" && x_active_tree == :ot_tree && @record && @record.in_use?
+      bgi[:items].each do |bgsi|
+        if bgsi[:button] == "orchestration_template_edit"
+          bgsi[:enabled] = 'false'
+          bgsi[:title] = _('Orchestration Templates that are in use cannot be edited')
+        elsif bgsi[:button] == "orchestration_template_remove"
+          bgsi[:enabled] = 'false'
+          bgsi[:title] = _('Orchestration Templates that are in use cannot be removed')
+        end
+      end
+    end
 
-    toolbar = []
-    groups_added = []
-    sep_needed = false
+    current_item = props
+    current_item[:items] ||= []
+    any_visible = false
+    bgi[:items].each_with_index do |bsi, bsi_idx|
+      if bsi.key?(:separator)
+        props = {"id" => "sep_#{index}_#{bsi_idx}", "type" => "separator", :hidden => !any_visible}
+      else
+        return nil if download_pdf_buttons.include?(bsi[:button]) && !PdfGenerator.available?
```

Comment:
used to be `next` here, which would jump out `bgi[:items].each_with_index`, but now the return actually nexts `      bg[:items].each do |bgi|` .. bug or a feature? :)